### PR TITLE
Add jonavin user space / common functions

### PR DESCRIPTION
**Pull request by Jonavin**
## Description

create jonavin userspace

consolidate common jonavin keymap functions into userspace

ENABLE FEATURES your keymap rules.mk

STARTUP_NUMLOCK_ON = yes
    turns on NUMLOCK by default

ENCODER_DEFAULTACTIONS_ENABLE = yes
    Enabled default encoder funtions
  
TD_LSFT_CAPSLOCK_ENABLE = yes
    This will enable double tap on Left Shift to toggle CAPSLOCK

IDLE_TIMEOUT_ENABLE = yes
    Enables Timer functionality; for RGB idle timeouts that can be changed dynamically
    
    Functions:
        u16int_t get_timeout_threshold(void)            // returns the current timeout threshold
        void timeout_update_threshold(bool increase)  // change threshold: true = increase, false = decrease     
        void timeout_reset_timer(void)  // resets timer (put in process_record_user if you override it)
        void timeout_tick_timer(void)   // registers time ticks (put in maxtrix_scan_user if you override it)

Other Functions:
activate_numlock(bool turn_on)  // true = turn on NUM LOCK, false = off 

BUG FIX gmmk pro keymap:  RSFT+encoder not working as intended due to register/unreg of wrong Shift mod.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* potentially addresses issue #13580

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: users/jonavin/jonavin.c [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                if (timeout_threshold > 0 && timeout_counter >= timeout_threshold) {
+                    rgb_matrix_disable_noeeprom();
+                }
+            #endif
+        } // timeout_threshold = 0 will disable timeout
+    }
+
+    __attribute__ ((weak))  void matrix_scan_user(void) {
+        timeout_tick_timer();
+    };
```
Comment (drashna) [inline users/jonavin/jonavin.c:103]: Something like this may be better: 
```suggestion
__attribute__((weak)) void matrix_scan_keymap(void) {}

void matrix_scan_user(void) {
    timeout_tick_timer();
    matrix_scan_keymap();
}
```

File: users/jonavin/jonavin.c [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+    #ifdef RGB_MATRIX_ENABLE
+        __attribute__ ((weak)) void suspend_power_down_user(void) {
+            rgb_matrix_set_suspend_state(true);
+        }
+
+        __attribute__ ((weak)) void suspend_wakeup_init_user(void) {
+            rgb_matrix_set_suspend_state(false);
+        }
+    #endif // RGB_MATRIX_ENABLE
```
Comment (drashna) [inline users/jonavin/jonavin.c:113]: This is not needed.
```suggestion
```

File: users/jonavin/jonavin.c [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+
+#endif // IDLE_TIMEOUT_ENABLE
+
+
+#ifdef ENCODER_ENABLE       // Encoder Functionality
+    #ifndef DYNAMIC_KEYMAP_LAYER_COUNT
+        #define DYNAMIC_KEYMAP_LAYER_COUNT 4  //default in case this is not already defined elsewhere
+    #endif
+    __attribute__ ((weak)) uint8_t selected_layer = 0;
```
Comment (drashna) [inline users/jonavin/jonavin.c:122]:
```suggestion
uint8_t selected_layer = 0;
```

File: users/jonavin/jonavin.c [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+    __attribute__ ((weak)) bool encoder_update_user(uint8_t index, bool clockwise) {
```
Comment (drashna) [inline users/jonavin/jonavin.c:124]:
```suggestion
__attribute__((weak)) bool encoder_update_keymap(uint8_t index, bool clockwise) { return true; }

bool encoder_update_user(uint8_t index, bool clockwise) {
    if (!encoder_update_keymap(index, clockwise)) { return false; }
```

File: users/jonavin/jonavin.c [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+  RGB_TOI,   // Timeout idle time up
+  RGB_TOD,   // Timeout idle time down
+};
+
+#define KC_CAD	LALT(LCTL(KC_DEL))
+#define KC_AF4	LALT(KC_F4)
+#define KC_TASK	LCTL(LSFT(KC_ESC))
+
+// Win Key Lock
+__attribute__ ((weak)) bool _isWinKeyDisabled = false;
```
[the remark this answers is not in the document]
Comment (Jonavin) [reply users/jonavin/jonavin.c:49]: I'll rework some of this to avoid breaking existing keymaps

[the remark this answers is not in the document]
Comment (Jonavin) [reply users/jonavin/jonavin.c]: I switched to keymap_config.no_gui.  Didn't know that was available. That simplifies things even more.

**Comment by Jonavin** [comment]
Some of the recommended changes break existing keymaps. These have been updated to work with the userspace.

File: keyboards/keebio/quefrency/keymaps/jonavin/keymap.c [hunk 6]
```diff
@@ -16,36 +16,7 @@
...
-
-qk_tap_dance_action_t tap_dance_actions[] = {
-  // Tap once for shift, twice for Caps Lock
-  [TD_LSFT_CAPSLOCK] = ACTION_TAP_DANCE_DOUBLE(KC_LSFT, KC_CAPS),
-};
-
-#define KC_LSFTCAPS TD(TD_LSFT_CAPSLOCK)
-
-bool _isWinKeyDisabled = false;
+#include <jonavin.h>
```
Comment (drashna) [inline keyboards/keebio/quefrency/keymaps/jonavin/keymap.c:19]:
```suggestion
#include "jonavin.h"
```

File: keyboards/gmmk/pro/keymaps/jonavin/keymap.c [hunk 7]
```diff
@@ -98,135 +69,19 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
...
-    default:
-        if (record->event.pressed) { //reset activity timer
-            #ifdef RGB_MATRIX_ENABLE
-                rgb_matrix_enable();
-            #endif
-            timeout_reset_timer();
-        }
-        break;
-    }
+bool process_record_keymap(uint16_t keycode, keyrecord_t *record) {
```
Comment (drashna) [inline keyboards/gmmk/pro/keymaps/jonavin/keymap.c:72]: can be removed.

File: keyboards/gmmk/pro/keymaps/jonavin/keymap.c [hunk 8]
```diff
@@ -98,135 +69,19 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
...
-            timeout_counter++;
-            timeout_timer = timer_read();
-        }
-        #ifdef RGB_MATRIX_ENABLE
-            if (timeout_threshold > 0 && timeout_counter >= timeout_threshold) {
-                rgb_matrix_disable_noeeprom();
-            }
-        #endif
-    } // timeout_threshold = 0 will disable timeout
+void matrix_scan_keymap(void) {
```
Comment (drashna) [inline keyboards/gmmk/pro/keymaps/jonavin/keymap.c:77]: Same

File: keyboards/gmmk/pro/keymaps/jonavin/keymap.c [hunk 9]
```diff
@@ -98,135 +69,19 @@ const uint16_t PROGMEM keymaps[][MATRIX_ROWS][MATRIX_COLS] = {
...
-                    timeout_update_threshold(false);
-                    break;
-                default:
-                    tap_code(KC_VOLD);
-                    break;
-                }
-            }
-        }
-
+    bool encoder_update_keymap(uint8_t index, bool clockwise) {
```
Comment (drashna) [inline keyboards/gmmk/pro/keymaps/jonavin/keymap.c:83]: Same

File: users/jonavin/jonavin.c [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include QMK_KEYBOARD_H
+#include <jonavin.h>
```
Comment (drashna) [inline users/jonavin/jonavin.c:20]:
```suggestion
#include "jonavin.h"
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/gmmk/pro/keymaps/jonavin/keymap.c (hunk 7 above)
Comment (Jonavin) [reply keyboards/gmmk/pro/keymaps/jonavin/keymap.c:72]: I added these as a placeholder and reminder for others who use my keymap or copy it.

Comment (Jonavin) [reply keyboards/gmmk/pro/keymaps/jonavin/keymap.c]: Moving info to readmes and cleaned up code.

File: .gitignore [hunk 11]
```diff
@@ -87,3 +87,4 @@ __pycache__
 
 # Allow to exist but don't include it in the repo
 user_song_list.h
+.vscode/settings.json
```
Comment (drashna) [inline .gitignore:90]: Could you revert this change?

File: .vscode/settings.json [hunk 12]
```diff
@@ -17,14 +17,15 @@
         "xstddef": "c",
         "type_traits": "c",
         "utility": "c",
-        "ranges": "c"
+        "ranges": "c",
```
Comment (drashna) [inline .vscode/settings.json:20]: Could you revert these changes?

**Review by drashna — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 11 above)
Comment (Jonavin) [reply .gitignore]: dond

File: .vscode/settings.json (hunk 12 above)
Comment (Jonavin) [reply .vscode/settings.json]: updated

**Review by fauxpark — APPROVED** [review]

File: .vscode/settings.json [hunk 13]
```diff
@@ -25,6 +25,6 @@
     },
     "python.formatting.provider": "yapf",
     "[json]": {
-	    "editor.formatOnSave": false
+        "editor.formatOnSave": false
```
Comment (drashna) [inline .vscode/settings.json:28]: Could this be reverted, as well?

Comment (fauxpark) [reply .vscode/settings.json:28]: Or at least done in a separate PR, as the change itself is good (fixing mixed tabs and spaces).

Comment (Jonavin) [reply .vscode/settings.json]: Updated

File: .vscode/settings.json [hunk 14]
```diff
@@ -25,6 +25,6 @@
     },
     "python.formatting.provider": "yapf",
     "[json]": {
-	    "editor.formatOnSave": false
+          "editor.formatOnSave": false
```
Comment (fauxpark) [inline .vscode/settings.json:28]:
```suggestion
	    "editor.formatOnSave": false
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!
